### DISCUSSION TITLE
Camera Controls

### DISCUSSION THREAD

**Discussion by MaWalla** [InfiniTimeOrg/InfiniTime#1600] [Ideas]
When having the phone on a tripod or in some other fixed position (for example for group photos, or astrophotography), it would be very convenient to have a shutter button on the PineTime. A nice addition (if applicable) would be the possibility of having manual controls such as ISO, exposure time, maybe even the focus there too.
The icing on the cake would be the ability to see the camera viewfinder on the watch, similar to what is possible with an apple watch, even if only at like 1 frame per second. I assume that both BLE will be too slow and resources (CPU and especially RAM) won't be enough for this to be possible (next to the implementation on the phone being rather difficult I guess).

For this, a seperate camera app within InfiniTime could be used to provide the necessary interface, but also an expansion to the companion app, which shouldn't be too difficult for the shutter button I think, but probably for the rest, if not straight up impossible there, but others may be more aware of the possibilities here than me.

Long story short: it would be awesome to have a shutter button in InfiniTime, more camera controls/viewfinder would make this even better.

**Comment by Avamander** [comment]
Streaming an image would in theory be doable. 

The bigger question is, are there any camera apps that are willing to receive external Intents for trigger and provide low-res viewfinder image trough a ContentProvider?

**Comment by MaWalla** [comment]
This is great to know! (I have another unrelated idea I'll put into another issue)

Anyway, the question is indeed good. I don't know if it's related, but maybe you know these selfie sticks which used a 3.5mm audio jack cable with a button near the handle to trigger the shutter. Since those are a thing, there must be some kind of interface to achieve it. I don't know if this can be translated to the companion app using it though.

**Comment by Avamander** [comment]
The audio jack ones tend to be TRRS, and emulate a headset button, it's usually not available to software.

**Comment by Landrash** [comment]
Cheap Bluetooth controllers usually send **Volume Up** and I've used them with a bunch of applications on Android without a problem.

**Comment by jmaris** [comment]
Sending volume up would likely be the best approach to this!

**Comment by 88572** [comment]
If it helps, here are the Google Camera defaults:
![Screenshot_20230729-073735.jpg](https://github.com/InfiniTimeOrg/InfiniTime/assets/82726593/7c4ce833-e5d7-4a7c-b6bd-82115e6e5aa7)

For OnePlus, it's the same. I can't check my old phone or my tablet to see what it is for One UI.

**Comment by jmlich** [comment]
The harbour-amazfish can handle custom events upon button pressed. It basically starts shell script which can do anything.

It implemented for miband service here:
https://github.com/piggz/harbour-amazfish/blob/master/daemon/src/services/mibandservice.cpp#L116
Its UUID_CHARACTERISTIC_MIBAND_DEVICE_EVENT and first byte of payload is compared with EVENT_BUTTON

Making script to control camera is another question, but IMHO it is doable.

for pinetime it could be implemented here:
https://github.com/piggz/harbour-amazfish/blob/11a9ea5e52e8afb697f3766fd1b7ad98095d72aa/daemon/src/devices/pinetimejfdevice.cpp#L415

In infinitime the code would be similar to AcceptIncommingCall
https://github.com/InfiniTimeOrg/InfiniTime/blob/46b664b5289f7ce7c4170d6d8a65cfd770dc855b/src/components/ble/AlertNotificationService.cpp#L88

However, I am confused how are those UUID constants created and if payload can be freely defined or it is according to some standard.
